### PR TITLE
MQE: Only loop through float points once

### PR DESCRIPTION
**Pull request by jhesketh**
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by charleskorn — COMMENTED** [review]
I'm a little suspicious of the benchmark results here - why have the results for cases that don't use `sum` (eg. `a_2000,_range_query_with_1000_steps`) seen such a degradation in performance? They should be unaffected.

**Review by charleskorn — COMMENTED** [review]
I would expect the old version of this to be more performant.

Let's say each group has 100 input series with 200 samples each.

If we use the change in this PR, we'll execute this `if` 100x200=20000 times:

```go
if !thisSeriesGroup.floatPresent[idx] {
	...
}
``` 

If we iterate over `floatPresent` once after the group has been constructed, then this loop will only have 200 iterations:

```go
pointCount := 0
for _, p := range thisGroup.floatPresent {
	if p {
		...
	}
}
```

**Comment by jhesketh** [comment]
I don't trust my benchmarks. Charles reran them:

<details>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/streamingpromql/benchmarks
                                                                          │   main.txt   │            pr-8442.txt             │
                                                                          │    sec/op    │   sec/op     vs base               │
Query/a_1,_instant_query/Mimir-10                                           133.5µ ± 11%   118.7µ ± 2%  -11.05% (p=0.002 n=6)
Query/a_1,_range_query_with_100_steps/Mimir-10                              146.1µ ±  9%   129.8µ ± 6%  -11.15% (p=0.002 n=6)
Query/a_1,_range_query_with_1000_steps/Mimir-10                             215.4µ ±  2%   202.8µ ± 2%   -5.87% (p=0.002 n=6)
Query/a_100,_instant_query/Mimir-10                                         841.0µ ±  1%   841.1µ ± 2%        ~ (p=0.937 n=6)
Query/a_100,_range_query_with_100_steps/Mimir-10                            1.444m ±  1%   1.442m ± 1%        ~ (p=0.485 n=6)
Query/a_100,_range_query_with_1000_steps/Mimir-10                           6.853m ±  4%   6.708m ± 1%   -2.12% (p=0.009 n=6)
Query/a_2000,_instant_query/Mimir-10                                        11.37m ±  1%   11.43m ± 1%        ~ (p=0.132 n=6)
Query/a_2000,_range_query_with_100_steps/Mimir-10                           22.14m ±  2%   22.15m ± 1%        ~ (p=0.699 n=6)
Query/a_2000,_range_query_with_1000_steps/Mimir-10                          117.1m ±  1%   117.5m ± 0%        ~ (p=0.310 n=6)
Query/a_1[1m],_instant_query/Mimir-10                                       110.2µ ±  1%   111.3µ ± 3%        ~ (p=0.394 n=6)
Query/a_100[1m],_instant_query/Mimir-10                                     370.8µ ±  1%   366.2µ ± 1%   -1.25% (p=0.004 n=6)
Query/a_2000[1m],_instant_query/Mimir-10                                    4.182m ±  0%   4.176m ± 1%        ~ (p=0.240 n=6)
Query/sum(a_1),_instant_query/Mimir-10                                      120.7µ ±  1%   122.2µ ± 3%        ~ (p=0.310 n=6)
Query/sum(a_1),_range_query_with_100_steps/Mimir-10                         131.5µ ±  1%   130.9µ ± 1%        ~ (p=0.240 n=6)
Query/sum(a_1),_range_query_with_1000_steps/Mimir-10                        208.2µ ±  1%   210.8µ ± 2%        ~ (p=0.240 n=6)
Query/sum(a_100),_instant_query/Mimir-10                                    839.0µ ±  2%   832.6µ ± 1%        ~ (p=0.132 n=6)
Query/sum(a_100),_range_query_with_100_steps/Mimir-10                       1.437m ±  1%   1.442m ± 2%        ~ (p=0.132 n=6)
Query/sum(a_100),_range_query_with_1000_steps/Mimir-10                      6.813m ±  2%   6.840m ± 1%        ~ (p=0.065 n=6)
Query/sum(a_2000),_instant_query/Mimir-10                                   11.39m ±  1%   11.40m ± 1%        ~ (p=0.937 n=6)
Query/sum(a_2000),_range_query_with_100_steps/Mimir-10                      22.22m ±  2%   22.44m ± 1%        ~ (p=0.132 n=6)
Query/sum(a_2000),_range_query_with_1000_steps/Mimir-10                     120.5m ±  1%   121.4m ± 0%   +0.73% (p=0.009 n=6)
Query/sum_by_(l)(h_1),_instant_query/Mimir-10                               166.8µ ±  1%   167.6µ ± 2%        ~ (p=0.132 n=6)
Query/sum_by_(l)(h_1),_range_query_with_100_steps/Mimir-10                  208.9µ ±  1%   207.2µ ± 1%        ~ (p=0.065 n=6)
Query/sum_by_(l)(h_1),_range_query_with_1000_steps/Mimir-10                 595.9µ ±  3%   593.1µ ± 2%   -0.46% (p=0.041 n=6)
Query/sum_by_(l)(h_100),_instant_query/Mimir-10                             4.054m ±  1%   4.053m ± 1%        ~ (p=0.937 n=6)
Query/sum_by_(l)(h_100),_range_query_with_100_steps/Mimir-10                7.534m ±  3%   7.553m ± 1%        ~ (p=0.818 n=6)
Query/sum_by_(l)(h_100),_range_query_with_1000_steps/Mimir-10               39.46m ±  1%   39.37m ± 0%        ~ (p=0.699 n=6)
Query/sum_by_(l)(h_2000),_instant_query/Mimir-10                            74.93m ±  1%   75.17m ± 1%        ~ (p=0.093 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_100_steps/Mimir-10               143.0m ±  5%   143.3m ± 1%        ~ (p=0.699 n=6)
Query/sum_by_(l)(h_2000),_range_query_with_1000_steps/Mimir-10              750.9m ±  2%   753.9m ± 1%        ~ (p=0.699 n=6)
Query/sum_by_(le)(h_1),_instant_query/Mimir-10                              173.5µ ±  7%   168.7µ ± 2%   -2.75% (p=0.026 n=6)
Query/sum_by_(le)(h_1),_range_query_with_100_steps/Mimir-10                 222.1µ ± 11%   213.5µ ± 4%        ~ (p=0.180 n=6)
Query/sum_by_(le)(h_1),_range_query_with_1000_steps/Mimir-10                620.8µ ±  3%   604.1µ ± 4%        ~ (p=0.180 n=6)
Query/sum_by_(le)(h_100),_instant_query/Mimir-10                            4.011m ±  7%   4.056m ± 0%        ~ (p=0.065 n=6)
Query/sum_by_(le)(h_100),_range_query_with_100_steps/Mimir-10               7.451m ±  2%   7.475m ± 1%        ~ (p=0.310 n=6)
Query/sum_by_(le)(h_100),_range_query_with_1000_steps/Mimir-10              39.02m ±  1%   39.26m ± 0%   +0.61% (p=0.015 n=6)
Query/sum_by_(le)(h_2000),_instant_query/Mimir-10                           73.83m ±  1%   73.48m ± 1%   -0.48% (p=0.041 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_100_steps/Mimir-10              141.1m ±  1%   141.5m ± 1%        ~ (p=0.240 n=6)
Query/sum_by_(le)(h_2000),_range_query_with_1000_steps/Mimir-10             742.2m ±  1%   754.3m ± 1%   +1.63% (p=0.002 n=6)
Query/sum(a_1_+_b_1),_instant_query/Mimir-10                                228.9µ ±  2%   229.4µ ± 1%        ~ (p=1.000 n=6)
Query/sum(a_1_+_b_1),_range_query_with_100_steps/Mimir-10                   248.0µ ±  1%   249.3µ ± 1%        ~ (p=0.065 n=6)
Query/sum(a_1_+_b_1),_range_query_with_1000_steps/Mimir-10                  392.3µ ±  2%   393.6µ ± 1%        ~ (p=0.589 n=6)
Query/sum(a_100_+_b_100),_instant_query/Mimir-10                            1.619m ±  3%   1.632m ± 1%        ~ (p=0.093 n=6)
Query/sum(a_100_+_b_100),_range_query_with_100_steps/Mimir-10               2.870m ±  5%   2.781m ± 1%   -3.09% (p=0.015 n=6)
Query/sum(a_100_+_b_100),_range_query_with_1000_steps/Mimir-10              13.31m ± 10%   13.35m ± 1%        ~ (p=0.589 n=6)
Query/sum(a_2000_+_b_2000),_instant_query/Mimir-10                          25.07m ±  1%   25.25m ± 1%        ~ (p=0.132 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_100_steps/Mimir-10             46.99m ±  3%   46.92m ± 2%        ~ (p=1.000 n=6)
Query/sum(a_2000_+_b_2000),_range_query_with_1000_steps/Mimir-10            245.5m ±  1%   247.3m ± 2%   +0.72% (p=0.015 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_instant_query/Mimir-10                    134.4µ ±  2%   135.3µ ± 1%        ~ (p=0.485 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_100_steps/Mimir-10       194.1µ ± 15%   190.1µ ± 2%   -2.07% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_1[1m])),_range_query_with_1000_steps/Mimir-10      639.5µ ±  3%   639.5µ ± 2%        ~ (p=0.818 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_instant_query/Mimir-10                  1.521m ±  0%   1.527m ± 1%        ~ (p=0.394 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_100_steps/Mimir-10     5.632m ±  1%   5.694m ± 1%   +1.11% (p=0.004 n=6)
Query/sum_by_(le)(rate(h_100[1m])),_range_query_with_1000_steps/Mimir-10    42.54m ±  1%   42.72m ± 0%   +0.41% (p=0.041 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_instant_query/Mimir-10                 26.06m ±  1%   26.29m ± 2%        ~ (p=0.132 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_100_steps/Mimir-10    106.0m ±  0%   106.9m ± 1%   +0.86% (p=0.002 n=6)
Query/sum_by_(le)(rate(h_2000[1m])),_range_query_with_1000_steps/Mimir-10   814.8m ±  1%   821.0m ± 2%   +0.76% (p=0.041 n=6)
geomean                                                                     4.036m         4.014m        -0.53%
```

</details>

> tl;dr: queries with many series and many steps get a little slower, queries with few get a little faster

We agreed that we should favour making bigger queries faster as they are slower in absolute terms.

I will add a comment to this effect.

**Review by krajorama — APPROVED** [review]
stamp

**Comment by dimitarvdimitrov** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!
